README.rst
PySAT: SAT technology in Python
===============================

PySAT is a Python (2.7, 3.4+) toolkit, which aims at providing a simple and
unified interface to a number of state-of-art `Boolean satisfiability (SAT)
<https://en.wikipedia.org/wiki/Boolean_satisfiability_problem>`__ solvers as
well as to a variety of cardinality and pseudo-Boolean encodings. The purpose
of PySAT is to enable researchers working on SAT and its applications and
generalizations to easily prototype with SAT oracles in Python while
exploiting incrementally the power of the original low-level implementations
of modern SAT solvers.

PySAT can be helpful when solving problems in `NP
<https://en.wikipedia.org/wiki/NP_(complexity)>`__ but also `beyond NP
<http://beyondnp.org/>`__. For instance, PySAT is handy when one needs to
quickly implement a MaxSAT solver, an MUS/MCS extractor or enumerator, an
abstraction-based QBF solver, or any other kind of tool solving an application
problem with the (potentially *multiple* and/or *incremental*) use of a SAT
oracle.

.. image:: web/medals.svg
   :width: 270 px
   :align: right

Features
--------

PySAT integrates a number of widely used state-of-the-art SAT solvers. All the
provided solvers are the original low-level implementations installed along
with PySAT. Note that the solvers' source code *is not* a part of the
project's source tree and is downloaded and patched at every PySAT
installation. Note that originally the solvers' source code was not
distributed with PySAT, which resulted in sequence of download and patch
operations for each solver during each installation of PySAT. This, however,
causes serious issues in case of using a proxy. As a result and since version
0.1.6.dev1, PySAT includes the solvers' archive files in the distribution.

Currently, the following SAT solvers are supported (at this point, for
Minisat-based solvers only *core* versions are integrated):

-  CaDiCaL (`rel-1.0.3 <https://github.com/arminbiere/cadical>`__)
-  Glucose (`3.0 <http://www.labri.fr/perso/lsimon/glucose/>`__)
-  Glucose (`4.1 <http://www.labri.fr/perso/lsimon/glucose/>`__)
-  Lingeling (`bbc-9230380-160707 <http://fmv.jku.at/lingeling/>`__)
-  MapleLCMDistChronoBT (`SAT competition 2018 version <http://sat2018.forsyte.tuwien.ac.at/solvers/main_and_glucose_hack/>`__)
-  MapleCM (`SAT competition 2018 version <http://sat2018.forsyte.tuwien.ac.at/solvers/main_and_glucose_hack/>`__)
-  Maplesat (`MapleCOMSPS_LRB <https://sites.google.com/a/gsd.uwaterloo.ca/maplesat/>`__)
-  Mergesat (`3.0 <https://github.com/conp-solutions/mergesat>`__)
-  Minicard (`1.2 <https://github.com/liffiton/minicard>`__)
-  Minisat (`2.2 release <http://minisat.se/MiniSat.html>`__)
-  Minisat (`GitHub version <https://github.com/niklasso/minisat>`__)

In order to make SAT-based prototyping easier, PySAT integrates a variety of
cardinality encodings. All of them are implemented from scratch in C++. The
list of cardinality encodings included is the following:

-  pairwise [8]_
-  bitwise [8]_
-  sequential counters [9]_
-  sorting networks [4]_
-  cardinality networks [2]_
-  ladder/regular [1]_ [5]_
-  totalizer [3]_
-  modulo totalizer [7]_
-  iterative totalizer [6]_

.. [1] Carlos Ansótegui, Felip Manyà. *Mapping Problems with Finite-Domain
   Variables to Problems with Boolean Variables*. SAT (Selected Papers) 2004.
   pp. 1-15

.. [2] Roberto Asin, Robert Nieuwenhuis, Albert Oliveras,
   Enric Rodriguez-Carbonell. *Cardinality Networks and Their Applications*.
   SAT 2009. pp. 167-180

.. [3] Olivier Bailleux, Yacine Boufkhad. *Efficient CNF Encoding of Boolean
   Cardinality Constraints*. CP 2003. pp. 108-122

.. [4] Kenneth E. Batcher. *Sorting Networks and Their Applications*.
   AFIPS Spring Joint Computing Conference 1968. pp. 307-314

.. [5] Ian P. Gent, Peter Nightingale. *A New Encoding of Alldifferent Into
   SAT*. In International workshop on modelling and reformulating constraint
   satisfaction problems 2004. pp. 95-110

.. [6] Ruben Martins, Saurabh Joshi, Vasco M. Manquinho, Inês Lynce.
   *Incremental Cardinality Constraints for MaxSAT*. CP 2014. pp. 531-548

.. [7] Toru Ogawa, Yangyang Liu, Ryuzo Hasegawa, Miyuki Koshimura,
   Hiroshi Fujita. *Modulo Based CNF Encoding of Cardinality Constraints and
   Its Application to MaxSAT Solvers*. ICTAI 2013. pp. 9-17

.. [8] Steven David Prestwich. *CNF Encodings*. Handbook of Satisfiability.
   2009. pp. 75-97

.. [9] Carsten Sinz. *Towards an Optimal CNF Encoding of Boolean
   Cardinality Constraints*. CP 2005. pp. 827-831

Furthermore, PySAT supports a number of encodings of pseudo-Boolean
constraints listed below. This is done by exploiting a third-party library
`PyPBLib <https://pypi.org/project/pypblib/>`__  developed by the `Logic
Optimization Group <http://ulog.udl.cat/>`__ of the University of Lleida.
(PyPBLib is a wrapper over the known PBLib library [10]_.)

-  binary decision diagrams (BDD) [11]_ [12]_
-  sequential weight counters [13]_
-  sorting networks [11]_
-  adder networks [11]_
-  and binary merge [14]_

.. [10] Tobias Philipp, Peter Steinke. *PBLib - A Library for Encoding
    Pseudo-Boolean Constraints into CNF*. SAT 2015. pp. 9-16

.. [11] Niklas Eén, Niklas Sörensson. *Translating Pseudo-Boolean
    Constraints into SAT*. JSAT. vol. 2(1-4). 2006. pp. 1-26

.. [12] Ignasi Abío, Robert Nieuwenhuis, Albert Oliveras,
    Enric Rodríguez-Carbonell. *BDDs for Pseudo-Boolean Constraints -
    Revisited*. SAT. 2011. pp. 61-75

.. [13] Steffen Hölldobler, Norbert Manthey, Peter Steinke. *A Compact
    Encoding of Pseudo-Boolean Constraints into SAT*. KI. 2012.
    pp. 107-118

.. [14] Norbert Manthey, Tobias Philipp, Peter Steinke. *A More Compact
    Translation of Pseudo-Boolean Constraints into CNF Such That
    Generalized Arc Consistency Is Maintained*. KI. 2014. pp. 123-134

Usage
-----

Boolean variables in PySAT are represented as natural identifiers, e.g. numbers
from :math:`\mathbb{N}_{>0}`. A *literal* in PySAT is assumed to be an integer,
e.g. ``-1`` represents a literal :math:`\neg{x_1}` while :math:`5` represents a
literal :math:`x_5`.  A *clause* is a list of literals, e.g. ``[-3, -2]`` is a
clause :math:`(\neg{x_3} \vee \neg{x_2})`.

The following is a trivial example of PySAT usage:

.. code:: python

    >>> from pysat.solvers import Glucose3
    >>>
    >>> g = Glucose3()
    >>> g.add_clause([-1, 2])
    >>> g.add_clause([-2, 3])
    >>> print(g.solve())
    >>> print(g.get_model())
    ...
    True
    [-1, -2, -3]

Another example shows how to extract *unsatisfiable cores* from a SAT
solver given an unsatisfiable set of clauses:

.. code:: python

    >>> from pysat.solvers import Minisat22
    >>>
    >>> with Minisat22(bootstrap_with=[[-1, 2], [-2, 3]]) as m:
    ...     print(m.solve(assumptions=[1, -3]))
    ...     print(m.get_core())
    ...
    False
    [-3, 1]

Finally, the following example gives an idea of how one can extract a
*proof* (supported by Glucose3, Glucose4, and Lingeling only):

.. code:: python

    >>> from pysat.formula import CNF
    >>> from pysat.solvers import Lingeling
    >>>
    >>> formula = CNF()
    >>> formula.append([-1, 2])
    >>> formula.append([1, -2])
    >>> formula.append([-1, -2])
    >>> formula.append([1, 2])
    >>>
    >>> with Lingeling(bootstrap_with=formula.clauses, with_proof=True) as l:
    ...     if l.solve() == False:
    ...         print(l.get_proof())
    ...
    ['2 0', '1 0', '0']

PySAT usage is detailed in the `provided examples
<https://github.com/pysathq/pysat/tree/master/examples>`__. For instance, one
can find simple PySAT-based implementations of

-  Fu&Malik algorithm for MaxSAT [15]_
-  RC2/OLLITI algorithm for MaxSAT [19]_ [20]_
-  CLD-like algorithm for MCS extraction and enumeration [17]_
-  LBX-like algorithm for MCS extraction and enumeration [18]_
-  Deletion-based MUS extraction [16]_

.. [15] Zhaohui Fu, Sharad Malik. *On Solving the Partial MAX-SAT Problem*.
   SAT 2006. pp. 252-265

.. [16] Joao Marques Silva. *Minimal Unsatisfiability: Models, Algorithms and
   Applications*. ISMVL 2010. pp. 9-14

.. [17] Joao Marques-Silva, Federico Heras, Mikolas Janota, Alessandro Previti,
   Anton Belov. *On Computing Minimal Correction Subsets*. IJCAI 2013. pp.
   615-622

.. [18] Carlos Mencia, Alessandro Previti, Joao Marques-Silva. *Literal-Based
   MCS Extraction*. IJCAI 2015. pp. 1973-1979

.. [19] António Morgado, Carmine Dodaro, Joao Marques-Silva. *Core-Guided
   MaxSAT with Soft Cardinality Constraints*. CP 2014. pp. 564-573

.. [20] António Morgado, Alexey Ignatiev, Joao Marques-Silva. *MSCG: Robust
   Core-Guided MaxSAT Solving. System Description*. JSAT 2015. vol. 9,
   pp. 129-134

The examples are installed with PySAT as a subpackage and, thus, they can be
accessed internally in Python:

.. code:: python

    >>> from pysat.formula import CNF
    >>> from pysat.examples.lbx import LBX
    >>>
    >>> formula = CNF(from_file='input.cnf')
    >>> mcsls = LBX(formula)
    >>>
    >>> for mcs in mcsls.enumerate():
    ...     print(mcs)

Alternatively, they can be used as standalone executables, e.g. like this:

::

   $ lbx.py -e all -d -s g4 -v another-input.wcnf

Installation
------------

There are several ways to install PySAT. At this point, either way assumes you
are using a POSIX-compliant operating system with GNU `make
<https://www.gnu.org/software/make/>`__ and `patch
<http://savannah.gnu.org/projects/patch/>`__ installed and available from the
command line. Installation also relies on a C/C++ compiler supporting C++11,
e.g. `GCC <https://gcc.gnu.org/>`__ or `Clang <https://clang.llvm.org/>`__, as
well as the ``six`` `Python package <https://pypi.org/project/six/>`__.
Finally, in order to compile "C extensions" included as modules, the installer
requires the headers of `Python <https://www.python.org/>`__ and `zlib
<https://www.zlib.net/>`__. Both can be installed using the standard package
repositories.

Note that although version `0.1.5.dev1` of PySAT brings Microsoft Windows
support, the toolkit was not extensively tested on this system. If you find
out that something is broken on Windows, please, `let us know
<https://github.com/pysathq/pysat/issues>`__. Your input is important.

Also note that using Clang is preferred on MacOS as there may be an issue with
GCC *being unaware of* the command-line option ``--stdlib=libc++``. Clang is
available on MacOS by default. To enforce the installer to use it, you need to
set the environment variable ``CC`` to ``/usr/bin/clang``. For that, do
``export CC=/usr/bin/clang`` if using Bash, or ``setenv CC /usr/bin/clang`` if
using tsch. *This is not needed on Linux!*

Once all the prerequisites are installed, the simplest way to get and start
using PySAT is to install the latest stable release of the toolkit from `PyPI
<https://pypi.org/>`__:

::

    $ pip install python-sat[pblib,aiger]

We encourage you to install the *optional* dependencies `pblib` and `aiger`,
as in the previous command. However, if it cannot be done (e.g. if their
installation fails), you can install PySAT with the functionality of `aiger`
and `pblib` disabled:

::

    $ pip install python-sat

Once installed from PyPI, the toolkit at a later stage can be updated in the
following way:

::

    $ pip install -U python-sat

.. note::

    For some shells, e.g. *zsh*, you may need to put the package names into
    single quotes, i.e. use ``pip install 'python-sat[pblib,aiger]'``.

Alternatively, one can clone `the repository
<https://github.com/pysathq/pysat>`__ and execute the following command in the
local copy:

::

    $ python setup.py install

This will install the toolkit into the system's Python path. If another
destination directory is preferred, it can be set by

::

    $ python setup.py install --prefix=<where-to-install>

Both options (i.e. via ``pip`` or ``setup.py``) are supposed to download
and compile all the supported SAT solvers as well as prepare the
installation of PySAT.

Citation
--------

If PySAT has been significant to a project that leads to an academic
publication, please, acknowledge that fact by citing PySAT:

::

    @inproceedings{imms-sat18,
      author    = {Alexey Ignatiev and
                   Antonio Morgado and
                   Joao Marques{-}Silva},
      title     = {{PySAT:} {A} {Python} Toolkit for Prototyping
                   with {SAT} Oracles},
      booktitle = {SAT},
      pages     = {428--437},
      year      = {2018},
      url       = {https://doi.org/10.1007/978-3-319-94144-8_26},
      doi       = {10.1007/978-3-319-94144-8_26}
    }

To-Do
-----

PySAT toolkit is a work in progress. Although it can already be helpful in many

web/news.rst
====
News
====

PySAT is being developed with the *rolling release* model, which means we do
not deliver major releases from time to time. Instead, many small (sometimes
tiny) and frequent updates are provided. As a result, `upgrading
<installation.html>`_ your installation of the toolkit every now to keep it up
to date and then is good idea.

Changelog and more
------------------

03.06.2022 (*0.1.7.dev17*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Addition of AtMostK constraints on the fly is added to RC2, LBX and MCSls.
- Added support for additional hard constraints in Hitman, both in the form of
  clauses and AtMostK constraints.
- Added the eMUS mode to OptUx, i.e. it can enumerate subset-minimal MUSes.
- Added support for additional formula to be used in OptUx for stopping MUS
  enumeration as soon as the additional formula is covered by the set of
  MUSes.

22.02.2022 (*0.1.7.dev16*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a minor issue in RC2 related to making useless iterations in core
  minimization.
- Added support for WCNF+ formulas in OptUx.

04.12.2021 (*0.1.7.dev15*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a minor bug in stratified RC2 related to model enumeration for empty
  formulas.
- Updated Hitman so that it opts for plain RC2 if all weights are the same.
- Fixed cost MUS computation in OptUx.

24.11.2021 (*0.1.7.dev14*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Another attempt to fix the bug in topw handling if unspecified.

24.11.2021 (*0.1.7.dev13*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a minor issue in fm.py related to handling native AM1 constraints.
- Fixed a bug in WCNF parser, which caused incorrect top weight computation if
  not specified in the preamble of the file explicitly.

09.11.2021 (*0.1.7.dev12*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a bug related to handling CNF formulas and improper updating IDPool
  for PB constraints in the case of no auxiliary variables.

12.10.2021 (*0.1.7.dev11*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a bug related to meaningless bounds in cardinality constraints (both
  AtMost and AtLeast).

17.08.2021 (*0.1.7.dev10*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added cluster-based stratification to :class:`RC2`.
- Made the selection of stratification strategy in :class:`RC2` optional.

02.08.2021 (*0.1.7.dev9*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added proper handling of border cases when creating cardinality constraints.
  Done for all encodings, and both AtMostK and AtLeastK constraints.
- Added an ``UnsupportedBound`` exception to be raised whenever *pairwise*,
  *bitwise*, or *ladder* encodings are created with the bound in :math:`(1, N
  - 1)`.

30.07.2021 (*0.1.7.dev8*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a bug in :class:`IDPool` introduced in ``0.1.7.dev7``, related to
  ``None`` handling.

28.07.2021 (*0.1.7.dev7*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Updated :class:`IDPool` to return a new variable ID without an object.
- Updated :class:`RC2` to support IDPool instead of `self.topv`.

17.07.2021 (*0.1.7.dev6*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Modified the implemention of the am1 heuristic in :class:`RC2`.

26.06.2021 (*0.1.7.dev5*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Got rid of relaxation variables in :class:`RC2` - minor issue, which might
  still have affected the performance.

11.06.2021 (*0.1.7.dev4*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed compilation issues occurring for cadical on macOS >= 11.3 with the
  most recent version of clang. Thanks to András Salamon.

24.05.2021 (*0.1.7.dev3*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added a warning in `enum_models()` regarding unsatisfiability of the formula
  after model enumeration is finished.
- Fixed a couple of bugs in :class:`FM`.
- Similar changes in other example implementations.

20.04.2021 (*0.1.7.dev2*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Improved the sequential counters encoding based on Donald Knuth's
  irredundant variant, following the suggestion of Alex Healy.

31.03.2021 (*0.1.7.dev1*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- An attempt to fix compilation issues of Mergesat.

31.03.2021 (*0.1.7.dev0*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added basic interface to :class:`Mergesat3`.
- Ported :class:`Minicard`'s native cardinality support to Glucose, which
  resulted in solvers :class:`Gluecard3` and :class:`Gluecard4`.
- Added translation of :class:`Cadical`'s binary DRUP format to text-based.

25.03.2021 (*0.1.6.dev16*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- A few minor corrections in :class:`RC2`, :class:`RC2Stratified`, and
  :class:`Hitman` addressing the issues related to empty formulas.

23.03.2021 (*0.1.6.dev15*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- A minor correction in :class:`OptUx` related to MUS cost value.

23.03.2021 (*0.1.6.dev14*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- A minor correction in the ``setup.py`` script.

23.03.2021 (*0.1.6.dev13*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added a way to bootstrap :class:`.LBX` and :class:`.MCSls` for computing
  MCSes that include a specified subset of clauses.
- Updated :class:`Hitman` to support *weighted* hitting set problems.
- Added :class:`.OptUx`, which is an smallest/optimal MUS extractor and
  enumerator that aims to replicate the performance of Forqes.

13.02.2021 (*0.1.6.dev12*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed formula dumping in other formats (see :meth:`to_alien`).

24.11.2020 (*0.1.6.dev11*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added formula dumping in SMT-LIB2 (see :meth:`to_alien`).

19.11.2020 (*0.1.6.dev10*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed the number of propagations ins MapleChrono.

27.09.2020 (*0.1.6.dev9*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a bug in copying of :class:`CNFPlus` and :class:`WCNFPlus` objects.
- Added an attempt to write down OPB and LP formats (likely to be buggy!).

26.09.2020 (*0.1.6.dev8*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Fixed a bug in `solvers.py` related to the :class:`CNFPlus` iterator.

25.09.2020 (*0.1.6.dev7*)
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added an example for accum_stats().
- Fixed a minor bug in musx example.
- A few fixes in :mod:`pysat.formula`.

pysat/pb.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-
##
## pb.py
##
##  Created on: Mar 13, 2019
##      Author: Alexey S. Ignatiev
##      E-mail: aignatiev@ciencias.ulisboa.pt
##

"""
    ===============
    List of classes
    ===============

    .. autosummary::
        :nosignatures:

        EncType
        PBEnc

    ==================
    Module description
    ==================

    .. note::

        Functionality of this module is available only if the `PyPBLib`
        package is installed, e.g. from PyPI:

        .. code-block::

            $ pip install pypblib

    This module provides access to the basic functionality of the `PyPBLib
    library <https://pypi.org/project/pypblib/>`__ developed by the `Logic
    Optimization Group <http://ulog.udl.cat/>`__ of the University of Lleida.
    PyPBLib provides a user with an extensive Python API to the well-known
    `PBLib library <http://tools.computational-logic.org/content/pblib.php>`__
    [1]_. Note the PyPBLib has a number of `additional features
    <http://hardlog.udl.cat/static/doc/pypblib/html/index.html>`__ that cannot
    be accessed through PySAT *at this point*. (One concrete example is a
    range of cardinality encodings, which clash with the internal
    :mod:`pysat.card` module.) If a user needs some functionality of PyPBLib
    missing in this module, he/she may apply PyPBLib as a standalone library,
    when working with PySAT.

    .. [1] Tobias Philipp, Peter Steinke. *PBLib - A Library for Encoding
        Pseudo-Boolean Constraints into CNF*. SAT 2015. pp. 9-16

    A *pseudo-Boolean constraint* is a constraint of the form:
    :math:`\left(\sum_{i=1}^n{a_i\cdot x_i}\\right)\circ k`, where
    :math:`a_i\in\mathbb{N}`, :math:`x_i\in\{y_i,\\neg{y_i}\}`,
    :math:`y_i\in\mathbb{B}`, and :math:`\circ\in\{\leq,=,\geq\}`.
    Pseudo-Boolean constraints arise in a number of important practical
    applications. Thus, several *encodings* of pseudo-Boolean constraints into
    CNF formulas are known [2]_. The list of pseudo-Boolean encodings
    supported by this module include BDD [3]_ [4]_, sequential weight counters
    [5]_, sorting networks [3]_, adder networks [3]_, and binary merge [6]_.
    Access to all cardinality encodings can be made through the main class of
    this module, which is :class:`.PBEnc`.

    .. [2] Olivier Roussel, Vasco M. Manquinho. *Pseudo-Boolean and
        Cardinality Constraints*. Handbook of Satisfiability. 2009.
        pp. 695-733

    .. [3] Niklas Eén, Niklas Sörensson. *Translating Pseudo-Boolean
        Constraints into SAT*. JSAT. vol. 2(1-4). 2006. pp. 1-26

    .. [4] Ignasi Abío, Robert Nieuwenhuis, Albert Oliveras,
        Enric Rodríguez-Carbonell. *BDDs for Pseudo-Boolean Constraints -
        Revisited*. SAT. 2011. pp. 61-75

    .. [5] Steffen Hölldobler, Norbert Manthey, Peter Steinke. *A Compact
        Encoding of Pseudo-Boolean Constraints into SAT*. KI. 2012.
        pp. 107-118

    .. [6] Norbert Manthey, Tobias Philipp, Peter Steinke. *A More Compact
        Translation of Pseudo-Boolean Constraints into CNF Such That
        Generalized Arc Consistency Is Maintained*. KI. 2014. pp. 123-134

    ==============
    Module details
    ==============
"""

#
#==============================================================================
import math
from pysat.formula import CNF

# checking whether or not pypblib is available and working as expected
pblib_present = True
try:
    from pypblib import pblib
except ImportError:
    pblib_present = False


#
#==============================================================================
class NoSuchEncodingError(Exception):
    """
        This exception is raised when creating an unknown LEQ, GEQ, or Equals
        constraint encoding.
    """

    pass


#
#==============================================================================
class EncType(object):
    """
        This class represents a C-like ``enum`` type for choosing the
        pseudo-Boolean encoding to use. The values denoting the encodings are:

        ::

            best       = 0
            bdd        = 1
            seqcounter = 2
            sortnetwrk = 3
            adder      = 4
            binmerge   = 5

        The desired encoding can be selected either directly by its integer
        identifier, e.g. ``2``, or by its alphabetical name, e.g.
        ``EncType.seqcounter``.

        All the encodings are produced and returned as a list of clauses in
        the :class:`pysat.formula.CNF` format.

        Note that the encoding type can be set to ``best``, in which case the
        encoder selects one of the other encodings from the list (in most
        cases, this invokes the ``bdd`` encoder).
    """

    best       = 0
    bdd        = 1
    seqcounter = 2
    sortnetwrk = 3
    adder      = 4
    binmerge   = 5

    # mapping from internal encoding identifiers to the ones of PyPBLib
    _to_pbenc = {
            best:       pblib.PB_BEST,
            bdd:        pblib.PB_BDD,
            seqcounter: pblib.PB_SWC,
            sortnetwrk: pblib.PB_SORTINGNETWORKS,
            adder:      pblib.PB_ADDER,
            binmerge:   pblib.PB_BINARY_MERGE
        }

    # mapping from internal comparator identifiers to the ones of PyPBLib
    _to_pbcmp = {
            '<': pblib.LEQ,
            '>': pblib.GEQ,
            '=': pblib.BOTH
        }


#
#==============================================================================
class PBEnc(object):
    """
        Abstract class responsible for the creation of pseudo-Boolean
        constraints encoded to a CNF formula. The class has three main *class
        methods* for creating LEQ, GEQ, and Equals constraints. Given (1)
        either a list of weighted literals or a list of unweighted literals
        followed by a list of weights, (2) an integer bound and an encoding
        type, each of these methods returns an object of class
        :class:`pysat.formula.CNF` representing the resulting CNF formula.

        Since the class is abstract, there is no need to create an object of
        it. Instead, the methods should be called directly as class methods,
        e.g. ``PBEnc.atmost(wlits, bound)`` or ``PBEnc.equals(lits, weights,
        bound)``. An example usage is the following:

        .. code-block:: python

            >>> from pysat.pb import *
            >>> cnf = PBEnc.atmost(lits=[1, 2, 3], weights=[1, 2, 3], bound=3)
            >>> print(cnf.clauses)
            [[4], [-1, -5], [-2, -5], [5, -3, -6], [6]]
            >>> cnf = PBEnc.equals(lits=[1, 2, 3], weights=[1, 2, 3], bound=3, encoding=EncType.bdd)
            >>> print(cnf.clauses)
            [[4], [-5, -2], [-5, 2, -1], [-5, -1], [-6, 1], [-7, -2, 6], [-7, 2], [-7, 6], [-8, -3, 5], [-8, 3, 7], [-8, 5, 7], [8]]
    """

    @classmethod
    def _update_vids(cls, cnf, vpool):
        """
            Update variable ids in the given formula and id pool.

            :param cnf: a list of literals in the sum.
            :param vpool: the value of bound :math:`k`.

            :type cnf: :class:`.formula.CNF`
            :type vpool: :class:`.formula.IDPool`
        """

        top, vmap = vpool.top, {}  # current top and variable mapping

        # creating a new variable mapping, taking into
        # account variables marked as "occupied"
        while top < cnf.nv:
            top += 1
            vpool.top += 1

            while vpool._occupied and vpool.top >= vpool._occupied[0][0]:
                if vpool.top <= vpool._occupied[0][1] + 1:
                    vpool.top = vpool._occupied[0][1] + 1

                vpool._occupied.pop(0)

            vmap[top] = vpool.top

        # updating the clauses
        for cl in cnf.clauses:
            cl[:] = map(lambda l: int(math.copysign(vmap[abs(l)], l)) if abs(l) in vmap else l, cl)

        # updating the number of variables
        cnf.nv = vpool.top

    @classmethod
    def _encode(cls, lits, weights=None, bound=1, top_id=None, vpool=None,
            encoding=EncType.best, comparator='<'):
        """
            This is the method that wraps the encoder of PyPBLib. Although the
            method can be invoked directly, a user is expected to call one of
            the following methods instead: :meth:`atmost`, :meth:`atleast`, or
            :meth:`equals`.

            The list of literals can contain either integers or pairs ``(l,
            w)``, where ``l`` is an integer literal and ``w`` is an integer
            weight. The latter can be done only if no ``weights`` are
            specified separately.

            :param lits: a list of literals in the sum.
            :param weights: a list of weights
            :param bound: the value of bound :math:`k`.
            :param top_id: top variable identifier used so far.
            :param vpool: variable pool for counting the number of variables.
            :param encoding: identifier of the encoding to use.
            :param comparator: identifier of the comparison operator

            :type lits: iterable(int)
            :type weights: iterable(int)
            :type bound: int
            :type top_id: integer or None
            :type vpool: :class:`.IDPool`
            :type encoding: integer
            :type comparator: str

            :rtype: :class:`pysat.formula.CNF`
        """

        assert pblib_present, 'Package \'pypblib\' is unavailable. Check your installation.'

        if encoding < 0 or encoding > 5:
            raise(NoSuchEncodingError(encoding))

        assert lits, 'No literals are provided.'

        assert not top_id or not vpool, \
                'Use either a top id or a pool of variables but not both.'

        # preparing weighted literals
        if weights:
            assert len(lits) == len(weights), 'Same number of literals and weights is expected.'
            wlits = [pblib.WeightedLit(l, w) for l, w in zip(lits, weights)]
        else:
            if all(map(lambda lw: (type(lw) in (list, tuple)) and len(lw) == 2, lits)):
                # literals are already weighted
                wlits = [pblib.WeightedLit(*wl) for wl in lits]
                lits = zip(*lits)[0]  # unweighted literals for getting top_id
            elif all(map(lambda l: type(l) is int, lits)):
                # no weights are provided => all weights are units
                wlits = [pblib.WeightedLit(l, 1) for l in lits]
            else:
                assert 0, 'Incorrect literals given.'

        # obtaining the top id from the variable pool
        if vpool:
            top_id = vpool.top

        if not top_id:
            top_id = max(map(lambda x: abs(x), lits))

        # pseudo-Boolean constraint and variable manager
        constr = pblib.PBConstraint(wlits, EncType._to_pbcmp[comparator], bound)
        varmgr = pblib.AuxVarManager(top_id + 1)

        # encoder configuration
        config = pblib.PBConfig()
        config.set_PB_Encoder(EncType._to_pbenc[encoding])

        # encoding
        result = pblib.VectorClauseDatabase(config)
        pb2cnf = pblib.Pb2cnf(config)
        pb2cnf.encode(constr, result, varmgr)

        # extracting clauses
        ret = CNF(from_clauses=result.get_clauses())
        ret.nv = max(ret.nv, top_id)  # needed if no auxiliary variable is used

        # updating vpool if necessary
        if vpool:
            if vpool._occupied and vpool.top <= vpool._occupied[0][0] <= ret.nv:
                cls._update_vids(ret, vpool)
            else:
                vpool.top = ret.nv - 1
                vpool._next()

        return ret

    @classmethod
    def leq(cls, lits, weights=None, bound=1, top_id=None, vpool=None,
            encoding=EncType.best):
        """
            This method can be used for creating a CNF encoding of a LEQ
            (weighted AtMostK) constraint, i.e. of
            :math:`\sum_{i=1}^{n}{a_i\cdot x_i}\leq k`. The resulting set of
            clauses is returned as an object of class :class:`.formula.CNF`.

            The input list of literals can contain either integers or pairs
            ``(l, w)``, where ``l`` is an integer literal and ``w`` is an
            integer weight. The latter can be done only if no ``weights`` are
            specified separately. The type of encoding to use can be specified
            using the ``encoding`` parameter. By default, it is set to
            ``EncType.best``, i.e. it is up to the PBLib encoder to choose the
            encoding type.

            :param lits: a list of literals in the sum.
            :param weights: a list of weights
            :param bound: the value of bound :math:`k`.
            :param top_id: top variable identifier used so far.
            :param vpool: variable pool for counting the number of variables.
            :param encoding: identifier of the encoding to use.

            :type lits: iterable(int)
            :type weights: iterable(int)
            :type bound: int
            :type top_id: integer or None
            :type vpool: :class:`.IDPool`
            :type encoding: integer

            :rtype: :class:`pysat.formula.CNF`
        """

        return cls._encode(lits, weights=weights, bound=bound, top_id=top_id,
                vpool=vpool, encoding=encoding, comparator='<')

    @classmethod
    def atmost(cls, lits, weights=None, bound=1, top_id=None, vpool=None,
            encoding=EncType.best):
        """
            A synonim for :meth:`PBEnc.leq`.
        """

        return cls.leq(lits, weights=weights, bound=bound, top_id=top_id,
                vpool=vpool, encoding=encoding)

    @classmethod
    def geq(cls, lits, weights=None, bound=1, top_id=None, vpool=None,
            encoding=EncType.best):
        """
            This method can be used for creating a CNF encoding of a GEQ
            (weighted AtLeastK) constraint, i.e. of
            :math:`\sum_{i=1}^{n}{a_i\cdot x_i}\geq k`. The method shares the
            arguments and the return type with method :meth:`PBEnc.leq`.
            Please, see it for details.
        """

        return cls._encode(lits, weights=weights, bound=bound, top_id=top_id,
                vpool=vpool, encoding=encoding, comparator='>')

    @classmethod
    def atleast(cls, lits, weights=None, bound=1, top_id=None, vpool=None,
            encoding=EncType.best):
        """
            A synonym for :meth:`PBEnc.geq`.
        """

        return cls.geq(lits, weights=weights, bound=bound, top_id=top_id,
                vpool=vpool, encoding=encoding)

    @classmethod
    def equals(cls, lits, weights=None, bound=1, top_id=None, vpool=None,
            encoding=EncType.best):
        """
            This method can be used for creating a CNF encoding of a weighted
            EqualsK constraint, i.e. of :math:`\sum_{i=1}^{n}{a_i\cdot x_i}=
            k`. The method shares the arguments and the return type with
            method :meth:`PBEnc.leq`. Please, see it for details.
        """

        return cls._encode(lits, weights=weights, bound=bound, top_id=top_id,
                vpool=vpool, encoding=encoding, comparator='=')
